Python-Functions-Decorators.md

```diff
@@ -1,9 +1,255 @@
 # Python Decorators
 
-[PEP 0318 - Decorators for Functions and Methods](https://www.python.org/dev/peps/pep-0318/)
 
-[Python Docs - Decorators](https://docs.python.org/3/whatsnew/2.4.html?highlight=decorator)
+Decorator essentially work as wrappers. They modify the behviour of the code before and after a target function execution, without the need to modify the function itself, augmenting the original functionality, thus decorating it.
 
-Now that we have covered basics of Python, let us apply Python in [Web Development](Web-Development-in-Python)!
+Before going in detail about decorators there are some concepts that should be clear. In Python, functions are objects and we can do a lot of useful stuff with them.
 
-[Previous](Python-Functions)
+> ### Assigning funtions to a variables:
+
+    def greet(name):
+      return "hello "+name
+      greet_someone = greet
+      print greet_someone("John")
+
+> Outputs: hello John
+
+> ### Defining funnctions inside other functions:
+	
+    def greet(name):
+    def get_message():
+        return "Hello "
+
+    result = get_message()+name
+    return result
+	print greet("John")
+
+> Outputs: Hello John
+
+> ### Functions can also be passed as parameters to other functions:
+	
+    def greet(name):
+    	return "Hello " + name 
+	def call_func(func):
+    	other_name = "John"
+    return func(other_name)  
+
+	print call_func(greet)
+
+> Outputs: Hello John
+
+> ### Functions can return other functions:
+	In other words, functions generating other functions.
+	
+    	def compose_greet_func():
+    		def get_message():
+        		return "Hello there!"
+
+    		return get_message
+		greet = compose_greet_func()
+		print greet()
+
+> Outputs: Hello there!
+
+> ### Inner functions have access to the enclosing scope
+More commonly known as a [losure](http://www.shutupandship.com/2012/01/python-closures-explained.html). A very powerful pattern that we will come across while building decorators. Another thing to note, Python only allows [read access to the outer scope](http://www.tech-thoughts-blog.com/2013/07/writing-closure-in-python.html) and not assignment. Notice how we modified the example above to read a "name" argument from the enclosing scope of the inner function and return the new function.
+
+
+	def compose_greet_func(name):
+    def get_message():
+        return "Hello there "+name+"!"
+
+    return get_message
+
+    greet = compose_greet_func("John")
+    print greet()
+
+> Outputs: Hello there John!
+
```

Comment:
Check our other Wiki articles with Python snippets. We use https://repl.it services to create runnable snippets and leave a link, so that the reader can easily play with them to learn better.